### PR TITLE
Feature/ptwo 212 typescript migration

### PR DESCRIPTION
**Pull request by pgoc**

### CODE REVIEW

File: exmg-paper-combobox.ts [hunk 1]
```diff
@@ -174,20 +195,29 @@ export class PaperComboboxElement extends LitElement {
 
   private ignoreFocus: boolean = false;
 
+  private isAnyItemActive: boolean = true;
+
+  private isElementInitialized: boolean = false;
+
   private observers = this.getObservers();
```
Comment (markoffk) [inline exmg-paper-combobox.ts]: type hint?

File: exmg-paper-combobox.ts [hunk 2]
```diff
@@ -174,20 +195,29 @@ export class PaperComboboxElement extends LitElement {
...
 
+  private keyDownBackspaceDebounce = debounce(200);
```
Comment (markoffk) [inline exmg-paper-combobox.ts]: type hint?

File: exmg-paper-combobox.ts [hunk 3]
```diff
@@ -196,10 +226,10 @@ export class PaperComboboxElement extends LitElement {
     };
   }
 
-  private executeObservers(changedProperties: GenericPropertyValues<keyof this | PrivateProps>): void {
-    Object.entries(this.observers).forEach(([key, cb]) => {
-      if (changedProperties.has(<any>key)) {
-        cb();
+  private executeObservers(changedProperties: ChangedProps): void {
+    Object.entries<any>(this.observers).forEach(([key, cb]) => {
```
Comment (markoffk) [inline exmg-paper-combobox.ts]: can it be something other than *any* or just remove *any* at all?

File: exmg-paper-combobox.ts [hunk 4]
```diff
@@ -231,7 +278,7 @@ export class PaperComboboxElement extends LitElement {
     }
 
     const id = this.getSelectedItemKey(this.selectedItem);
-    if (typeof id === 'undefined') {
+    if (typeof id === 'undefined' || id === -1) {
```
Comment (markoffk) [inline exmg-paper-combobox.ts]: are you sure "-1" is not valid value? I think when "-1" you have to also check for attr-for-selected=false

File: exmg-paper-combobox.ts [hunk 5]
```diff
@@ -247,17 +294,22 @@ export class PaperComboboxElement extends LitElement {
...
-  filterItems() {
+  filterItems(): boolean {
     const items: NodeListOf<HTMLElement> = this.querySelectorAll('paper-item, paper-icon-item');
     const hasFilterPhrase = !!this.inputValue && this.inputValue.length > 0;
+    const phrase = hasFilterPhrase ? this.inputValue.toLowerCase().trim() : '';
+    let isAnyItemActive = false;
 
     items.forEach(item => {
-      if (hasFilterPhrase && item.textContent &&  item.textContent.indexOf(this.inputValue || '') === -1) {
+      if (hasFilterPhrase && item.textContent &&  item.textContent.toLowerCase().indexOf(phrase) === -1) {
```
Comment (markoffk) [inline exmg-paper-combobox.ts:304]: I did not analyze it deeply, but you use hasFilterPhrase that is outside item's scope inside foreach? Are you sure there is everything alright with logic? maybe there is need to move it to 1 if outside foreach?
In my opinion, now it is harder to analyze what happens here

File: exmg-paper-combobox.ts [hunk 6]
```diff
@@ -401,29 +488,80 @@ export class PaperComboboxElement extends LitElement {
     if (this.autoValidate) {
       window.addEventListener('click', this.onClick);
     }
+
+    this.addEventListener('iron-resize', this.onIronResize);
```
Comment (markoffk) [inline exmg-paper-combobox.ts]: this or this.shadowRoot ?

File: exmg-paper-combobox.ts (hunk 1 above)
Comment (pgoc) [reply exmg-paper-combobox.ts]: done

File: exmg-paper-combobox.ts (hunk 2 above)
Comment (pgoc) [reply exmg-paper-combobox.ts]: done

File: exmg-paper-combobox.ts (hunk 3 above)
Comment (pgoc) [reply exmg-paper-combobox.ts]: done

File: exmg-paper-combobox.ts (hunk 4 above)
Comment (pgoc) [reply exmg-paper-combobox.ts]: It was fine check as values from attributes are always string. However added more precise check

File: exmg-paper-combobox.ts (hunk 5 above)
Comment (pgoc) [reply exmg-paper-combobox.ts:304]: It is on purpose, don't need to repeat check in loop `!!this.inputValue && this.inputValue.length > 0` i can determine it before looping. Then there is more readable and a bit but more performant

File: exmg-paper-combobox.ts (hunk 4 above)
Comment (markoffk) [inline exmg-paper-combobox.ts]: check for selected attribute

Comment (pgoc) [reply exmg-paper-combobox.ts]: done

**Review by markoffk — APPROVED** [review]

**Pull request merged**